src/devices/machine/74259.cpp

```diff
@@ -0,0 +1,391 @@
+//  inputs without changing enable state
+//-------------------------------------------------
+
+void addressable_latch_device::write_abcd(u8 a, bool d)
+{
+	m_address = a & 7;
+	m_data = d;
+
+	if (m_enable)
+	{
+		if (m_clear)
+			clear_outputs(u8(m_data) << m_address);
+		else
+			update_bit();
+	}
+}
+
+//-------------------------------------------------
+//  enable_w - handle enable input (active low)
+//-------------------------------------------------
+
+WRITE_LINE_MEMBER(addressable_latch_device::enable_w)
+{
+	m_enable = !state;
+	if (m_enable)
+		update_bit();
+	else if (m_clear)
+		clear_outputs(0);
+}
+
+//-------------------------------------------------
+//  update_bit - update one of the eight output
+//  lines with a new data bit
+//-------------------------------------------------
+
+void addressable_latch_device::update_bit()
+{
+	// first verify that the selected bit is actually changing
+	if (BIT(m_q, m_address) == m_data)
+		return;
+
+	if (!m_clear)
+	{
+		// update selected bit with new data
+		m_q = (m_q & ~(1 << m_address)) | (u8(m_data) << m_address);
+	}
+	else
+	{
+		// clear any other bit that was formerly set
+		clear_outputs(0);
+		m_q = u8(m_data) << m_address;
+	}
+
+	// update output line via callback
+	if (!m_q_out_cb[m_address].isnull())
+		m_q_out_cb[m_address](m_data);
+
+	// update parallel output
+	if (!m_parallel_out_cb.isnull())
+		m_parallel_out_cb(0, m_q, 1 << m_address);
+
+	// do some logging
+	if (LOG_ALL_WRITES || (LOG_UNDEFINED_WRITES && m_q_out_cb[m_address].isnull() && m_parallel_out_cb.isnull()))
+		logerror("Q%d %s at %s\n", m_address, m_data ? "set" : "cleared", machine().describe_context());
+}
+
+//-------------------------------------------------
+//  write_d0 - bus-triggered write handler using
+//  LSB of data (or CRUOUT on TMS99xx)
+//-------------------------------------------------
+
+WRITE8_MEMBER(addressable_latch_device::write_d0)
+{
+	if (LOG_MYSTERY_BITS && data != 0x00 && data != 0x01 && data != 0xff)
+		logerror("Mystery bits written to Q%d:%s%s%s%s%s%s%s\n",
+			offset,
+			BIT(data, 7) ? " D7" : "",
+			BIT(data, 6) ? " D6" : "",
+			BIT(data, 5) ? " D5" : "",
+			BIT(data, 4) ? " D4" : "",
+			BIT(data, 3) ? " D3" : "",
+			BIT(data, 2) ? " D2" : "",
+			BIT(data, 1) ? " D1" : "");
+
+	write_bit(offset, BIT(data, 0));
+}
+
+//-------------------------------------------------
+//  write_d7 - bus-triggered write handler using
+//  MSB of (8-bit) data
+//-------------------------------------------------
+
+WRITE8_MEMBER(addressable_latch_device::write_d7)
+{
+	if (LOG_MYSTERY_BITS && data != 0x00 && data != 0x80 && data != 0xff)
+		logerror("Mystery bits written to Q%d:%s%s%s%s%s%s%s\n",
+			offset,
+			BIT(data, 6) ? " D6" : "",
+			BIT(data, 5) ? " D5" : "",
+			BIT(data, 4) ? " D4" : "",
+			BIT(data, 3) ? " D3" : "",
+			BIT(data, 2) ? " D2" : "",
+			BIT(data, 1) ? " D1" : "",
+			BIT(data, 0) ? " D0" : "");
+
+	write_bit(offset, BIT(data, 7));
```

Comment:
Perhaps, but I'm more inclined to believe that the address map is the proper place for specifying whatever the interface between an address-data bus and a device happens to be.